### ISSUE TITLE
Rule proposal: unreachable-code

### ISSUE THREAD

**Issue by kkmuffme** [typescript-eslint/typescript-eslint#7063]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Typescript has https://www.typescriptlang.org/tsconfig#allowUnreachableCode to (dis)allow unreachable code, as it checks for it by default. However:
>This does not affect errors on the basis of code which appears to be unreachable due to type analysis.

So tons of unreachable code is never reported, which is a problem when refactoring code, since you end up with tons of legacy code that could have been removed long ago.

### Fail Cases

```typescript
type myType = {
    'hello': string
};

function foo( arg: myType ) {
    if ('foo' in arg) {
        // this is impossible and can never be reached
        return true;
    }

    if (!('hello' in arg)) {
        // this is impossible and can never be reached
        return false;
    }

    if ('hello' in arg) {
        return true;
    }

    // this can never be reached either
    console.log( 'impossible' );

    return false;
}
```

### Pass Cases

```typescript
type myType = {
    'hello': string
};

function foo( arg: myType ) {
    if ('hello' in arg) {
        console.log( 'hello' );
    }

    console.log( 'world' );
}
```

**Comment by bradzacher** [comment]
See my comment on your other issue https://github.com/typescript-eslint/typescript-eslint/issues/7064#issuecomment-1563622451

The "useless" checks aren't actually truly useless due to how TS object types work. These checks could actually execute at runtime so flagging them as useless isn't strictly correct - which is the same reason TS is cautious about flagging them.

I don't think it's a good idea for us to warn against classes of checks that can actually be called at runtime.

There are a number of cases that we could handle that are provably terminal based on types (eg for `if ('hello' in x) { return true}` the false branch is unreachable if `x` had a non-optional property `hello`), but I do know that the control flow analysis logic to detect these cases is far from trivial. I'm not sure if the value to work ratio is high enough given that the rule wouldn't catch very many cases at all in most codebases.

It would also likely be a bit of a moving target as we wouldn't want to duplicate reports that TS itself does - so we'd have to be careful.

As such I don't think such a rule belongs in this project. I'd like to hear thoughts from others though cc @typescript-eslint/triage-team

**Comment by JoshuaKGoldberg** [comment]
Yeah this sounds somewhat nice but _very_ difficult for us. I'd love to see you/someone try it out as an external plugin with our [Custom Rules docs](https://typescript-eslint.io/custom-rules).

-1 from me on including this in core until it's been tried out a bunch, but +1 to the idea of it existing in general. 😄
